File: backend/src/main/java/com/gobang/engine/rules/GameRules.java
package com.gobang.engine.rules;

import java.util.List;

public interface GameRules {
    String getName();

    boolean isValidMove(int[][] board, int row, int col, int player);

    boolean checkWin(int[][] board, int row, int col, int player);

    boolean isBoardFull(int[][] board);

    int evaluateBoard(int[][] board, int aiPlayer);

    List<int[]> getCandidateMoves(int[][] board);
}

File: frontend/src/game/engine/rules/GameRules.ts
import type { BoardState } from '../types';

export interface GameRules {
  readonly name: string;

  isValidMove(board: BoardState, row: number, col: number, player: number): boolean;

  checkWin(board: BoardState, row: number, col: number, player: number): boolean;

  isBoardFull(board: BoardState): boolean;

  evaluateBoard(board: BoardState, aiPlayer: number): number;

  getCandidateMoves(board: BoardState): [number, number][];
}
